src/components/gltf-model-plus.js

```diff
@@ -341,127 +340,221 @@ function runMigration(version, json) {
-  loadingManager.setURLModifier(getCustomGLTFParserURLResolver(gltfUrl));
-  const gltfLoader = new THREE.GLTFLoader(loadingManager);
+  beforeRoot() {
+    const parser = this.parser;
+    const jsonPreprocessor = this.jsonPreprocessor;
 
-  // TODO some models are loaded before the renderer exists. This is likely things like the camera tool and loading cube.
-  // They don't currently use KTX textures but if they did this would be an issue. Fixing this is hard but is part of
-  // "taking control of the render loop" which is something we want to tackle for many reasons.
-  if (!ktxLoader && AFRAME && AFRAME.scenes && AFRAME.scenes[0]) {
-    ktxLoader = new KTX2Loader(loadingManager).detectSupport(AFRAME.scenes[0].renderer);
-  }
+    //
+    if (jsonPreprocessor) {
+      parser.json = jsonPreprocessor(parser.json);
+    }
 
-  if (ktxLoader) {
-    gltfLoader.setKTX2Loader(ktxLoader);
-  }
+    //
+    let version = 0;
+    if (
+      parser.json.extensions &&
+      parser.json.extensions.MOZ_hubs_components &&
+      parser.json.extensions.MOZ_hubs_components.hasOwnProperty("version")
+    ) {
+      version = parser.json.extensions.MOZ_hubs_components.version;
+    }
+    runMigration(version, parser.json);
 
-  const parser = await new Promise((resolve, reject) => gltfLoader.createParser(gltfUrl, resolve, onProgress, reject));
+    // Note: Here may be rewritten with the one with parser.associations
+    const nodes = parser.json.nodes;
+    if (nodes) {
+      for (let i = 0; i < nodes.length; i++) {
+        const node = nodes[i];
 
-  parser.textureLoader = new HubsTextureLoader(loadingManager);
+        if (!node.extras) {
+          node.extras = {};
+        }
 
-  if (jsonPreprocessor) {
-    parser.json = jsonPreprocessor(parser.json);
+        node.extras.gltfIndex = i;
+      }
+    }
   }
 
-  let version = 0;
-  if (
-    parser.json.extensions &&
-    parser.json.extensions.MOZ_hubs_components &&
-    parser.json.extensions.MOZ_hubs_components.hasOwnProperty("version")
-  ) {
-    version = parser.json.extensions.MOZ_hubs_components.version;
-  }
-  runMigration(version, parser.json);
+  afterRoot(gltf) {
+    gltf.scene.traverse(object => {
+      // GLTFLoader sets matrixAutoUpdate on animated objects, we want to keep the defaults
+      // @TODO: Should this be fixed in the gltf loader?
+      object.matrixAutoUpdate = THREE.Object3D.DefaultMatrixAutoUpdate;
+      const materialQuality = window.APP.store.materialQualitySetting;
+      object.material = mapMaterials(object, material => convertStandardMaterial(material, materialQuality));
+    });
 
-  const nodes = parser.json.nodes;
-  if (nodes) {
-    for (let i = 0; i < nodes.length; i++) {
-      const node = nodes[i];
+    //
+    gltf.scene.traverse(object => {
+      if (!object.material) {
+        return;
+      }
+      const materials = Array.isArray(object.material) ? object.material : [object.material];
+      // @TODO: Handle more efficiently
```

Comment:
We are already overriding `parser.textureLoader` in the constructor. Seems like we can still use the HubsTextureLoader to take care of this no?

Reply:
I didn't really want to override `parser.textureLoader` because `GLTFLoader` doesn't have a proper API to override the texture loader. Overriding `parser.textureLoader = new CustomLoader()` is hacky and the `GLTFLoader` doesn't expect the texture loader to be overridden. There might be a chance that it causes future errors especially when updating to the future glTF loader.

I did override the texture loader because there doesn't seem to be other options to let `GLTFLoader` use `ImageBitmapLoader` for Firefox. Yes, overriding the texture loader is hacky as I wrote but `GLTFLoader` expects that the texture loader can be `ImageBitmapLoader` so I think it would be less problematic than using our custom texture loader. And if Firefox and Three.js will be updated to support `ImageBitmapLoader` we will be able to simply drop this hacky code.

But if we realize that this traverse is slow I want to revisit.